File: Project/SearchEngine.java

public class SearchEngine {

    private Invertedindex index;
    private LRUCache cache;
    private SpatialLoader loader;
    private String folderPath;
    private SessionStats stats = new SessionStats();

    public void initialize(String folderPath) {
        this.folderPath = folderPath;
        index = new Invertedindex(); // this creeating object of invertedIndex class
        index.buildIndex(folderPath); // this calling to build the index from invertedindex claass
        cache = new LRUCache(100);
        loader = new SpatialLoader(folderPath, index);
        System.out.println("Ready. " + index.size() + " terms indexed.");
    }

    public String search(String word) {
        word = word.trim().toLowerCase(); // this convert to lowercase and remove spaces
        System.out.println("Searching \n" + word + " ' ");
        if (cache.contains(word)) { // search in cache
            stats.recordHit();
            System.out.println("[CACHE HIT]");
            return cache.get(word);
        }
        stats.recordMiss();
        System.out.println("[CACHE MISS]");
        String fileName = index.getFileName(word);//this will get the file name in where the topic is present
        if (fileName == null) {
            System.out.println("Not found in any file");
            return null;
        }
        System.out.println("Found in file " + fileName);
        loader.loadNeighborsIntoCache(word, fileName, cache);
        // loader.loadNeighborsIntoCache(word, fileName, cache);

        String definition = cache.get(word);
        if (definition == null) {
            System.out.println("Failed to load: " + word);
            return null;
        }
        return definition;

    }

    public int getIndexSize() {
        return index.size();
    }

    public void printCacheState() {
        cache.printCacheState();
    }

    public SessionStats getStats() {
        return stats;
    }

    public String getSourceFile(String word) {
        return index.getFileName(word.toLowerCase().trim());
    }
}

File: backend/SearchEngine.js
const InvertedIndex = require('./InvertedIndex');
const LRUCache = require('./LRUCache');
const SpatialLoader = require('./SpatialLoader');
const SessionStats = require('./SessionStats');
const Trie = require('./TrieNode');

class SearchEngine {
    constructor() {
        this.index = null;
        this.cache = null;
        this.loader = null;
        this.stats = new SessionStats();
        this.trie = new Trie();
        this.folderPath = '';
    }

    initialize(folderPath) {
        this.folderPath = folderPath;
        this.index = new InvertedIndex();
        this.index.buildIndex(folderPath);
        this.cache = new LRUCache(100);
        this.loader = new SpatialLoader(folderPath, this.index);

        const allWords = this.index.getAllWords();
        for (const word of allWords) {
            this.trie.insert(word);
        }

        console.log(`Ready. ${this.index.size()} terms indexed.`);
    }

    search(word) {
        word = word.trim().toLowerCase();
        console.log(`Searching '${word}'`);

        if (this.cache.contains(word)) {
            this.stats.recordHit();
            console.log('[CACHE HIT]');
            return {
                definition: this.cache.get(word),
                source: this.index.getFileName(word),
                cacheHit: true
            };
        }

        this.stats.recordMiss();
        console.log('[CACHE MISS]');

        const fileName = this.index.getFileName(word);
        if (!fileName) {
            console.log('Not found in any file');
            return {
                definition: null,
                source: null,
                cacheHit: false
            };
        }

        console.log('Found in file', fileName);
        this.loader.loadNeighborsIntoCache(word, fileName, this.cache);

        const definition = this.cache.get(word);
        if (!definition) {
            console.log('Failed to load:', word);
            return {
                definition: null,
                source: fileName,
                cacheHit: false
            };
        }

        return {
            definition,
            source: fileName,
            cacheHit: false
        };
    }

    getIndexSize() {
        return this.index.size();
    }

    getCacheState() {
        return this.cache.getCacheState();
    }

    getStats() {
        return this.stats.getStats();
    }

    getSourceFile(word) {
        return this.index.getFileName(word.toLowerCase().trim());
    }

    autocomplete(prefix) {
        return this.trie.search(prefix);
    }

    addWord(word, definition, category) {
        const result = this.index.addWord(word, definition, category);
        if (result) {
            this.trie.insert(word);
        }
        return result;
    }

    getAllWords() {
        return this.index.getAllWords();
    }

    rebuildIndex() {
        this.initialize(this.folderPath);
    }
}

module.exports = SearchEngine;
